HotspotFinder
================

This repository contains HotspotFinder, a tool for the identification and annotation of hotspots of somatic mutations across the genome.

HotspotFinder first identifies recurrently mutated single nucleotide positions or hotspots across the genome using 
somatic mutations from cancer samples. This is, HotspotFinder finds which genomic positions are mutated in two or 
more patients. Then, HotspotFinder annotates hotspots with a set of genomic data to guide their interpretation. 


# License


HotspotFinder is available to the general public subject to certain conditions described in its [LICENSE](other_file.md)

# Installation

HotspotFinder is implemented in **Python 3.6**.

You can obtain the latest code from the repository and install it with pip:

    $ git clone git@bitbucket.org:bbglab/hotspotfinder.git
    $ cd hotspotfinder
    $ pip install .


The first time that you run HotspotFinder it will automatically download data from our servers. By default the
downloaded datasets go to ``~/.bgdata``. If you want to move these datasets to another folder you have to define the
system environment variable BGDATA_LOCAL with an export command.

The following command will show you the help:

    $ hotspotfinder --help

You can run HotspotFinder as: 

    $ hotspotfinder -i {input_file} -o {output_path} -conf {config_file}
 

# Input files and annotations

The basic HotspotFinder run only requires a file containing somatic mutations: single nucleotide variants (SNVs), 
multiple nucleotide variants (MNVs) and small insertions and deletions (indels). Genomic data will be automatically 
downloaded within the package installation. Additionally, users can provide their own paths to genomic annotations as 
described below. 

### Mutations
HotspotFinder's main input file is a TSV or CSV file containing SNVs, MNVs and short indels from a cancer cohort 
generated by whole genome (WGS), whole exome (WXS) or panel sequencing. In the current version, mutations are 
supported in the human genome build hg38. The mutations file requires the following header with the following 
information: 
 - CHROMOSOME: chromosome 1-22, X or Y. ('chr' prefix not included)
 - POSITION: genomic coordinate
 - REF: reference nucleotide(s)
 - ALT: alternate nucleotide(s)
 - SAMPLE: sample identifier

By default, hotspots are identified using all samples in the mutations file. Alternatively, HotspotFinder 
can report hotspots independently in sub-groups or categories drawn from the input file. 
This feature requires adding an extra column with one of these headers: 'GROUP', 'GROUP_BY', 'COHORT', 
'CANCER_TYPE', 'PLATFORM', and 'TYPE'. **The new column name must be specified in the configuration file.** 

### Genomic annotations
HotspotFinder incorporates genomic information to annotate hotspots and help in their interpretation. 
These annotations include: genome mappability, repeats and low complexity regions, population variants, 
immunoglobulin loci, genomic elements (coding sequences from protein coding transcripts, splice sites and introns, 
5’ and 3’ untranslated regions, promoters and other cis-regulatory regions, non-coding RNAs, and pseudogenes). See 
publication for further details. 

These files are automatically downloaded from bgdata the first time that HotspotFinder is run. 
Users can also replace the automatic annotations with their own data by adding paths to the files within 
the configuration file. If users do not want to annotate hotspots, they can disable the annotations in the 
configuration file. 

# Output files
HotspotFinder generates two output files: results.tsv and warningpositions.txt

### Hotspots results
The main output of HotspotFinder is saved in the TSV file **results.tsv**. Each row in the results file corresponds 
to a hotspot. Hotspots are independently identified in each of the four mutation types: 
single nucleotide variants (SNVs), multi-nucleotide variants (MNVs), short insertions and short deletions. 

The basic columns containing hotspots are:
- CHROMOSOME: chromosome
- POSITION: position
- CHR_POS: chromosome and position
- HOTSPOT_ID: unique hotspot identifier
- MUT_TYPE: mutation type
- COHORT: input cohort (file) name
- N_MUTATIONS: number of mutations in the hotspot
- N_MUTATED_SAMPLES: number of mutated samples in the hotspot
- FRAC_MUTATED_SAMPLES: fraction of mutated samples in the cohort
- REF: reference nucleotide
- ALT: alternate nucleotide(s)
- ALT_COUNTS: number of alternate nucleotide(s)
- FRAC_ALT: proportion of alternates among total alternates
- CONTEXT_3: trinucleotide context centred at the hotspot
- CONTEXT_5: pentanucleotide context centred at the hotspot
- N_COHORT_SAMPLES: number of samples in the cohort
- N_COHORT_MUTATIONS_TOTAL: number of total mutations in the cohort
- N_COHORT_MUTATIONS_SNV: number of single base substitutions in the cohort
- N_COHORT_MUTATIONS_MNV: number of multi-nucleotide substitutions in the cohort
- N_COHORT_MUTATIONS_INS: number of insertions in the cohort
- N_COHORT_MUTATIONS_DEL: number of deletions in the cohort
- MUTATED_SAMPLES: identifier of mutated samples
- MUTATED_SAMPLES_ALTS: identifier and corresponding alternate of mutated samples  
- OVERLAP_WARNING_POSITION: True if the hotspot overlaps a warning position

The additional columns for annotations are:  

- GENOMIC_ELEMENT: overlapping genomic element(s) 
- SYMBOL: symbols off overlapping genomic element(s)
- GENE_ID: Ensembl gene identifiers of overlapping genomic elements
- TRANSCRIPT_ID: Ensembl transcript identifiers of overlapping genomic elements
- GENOMIC_REGION: class of overlapping genomic element(s) 
- GENOMIC_REGION_PRIORITY: genomic element class with highest priority
- CODING_NONCODING: overlapping coding or non-coding genomic elements 
- MAPPABILITY: high or low according to GEM mappability regions overlap
- MAPPABILITY_BLACKLIST: pass if no overlap with blacklisted mappability regions
- VARIATION_AF: overlapping population variants
- HOTSPOTFINDER_FILTERS: flag if hotspot is mappable and does not overlap blacklisted mappability regions or population variants. 
- REPEATS: overlapping repeats
- REPEATS_OVERLAP: flag if hotspot overlaps repeats
- IG_TR: overlapping immunoglobulin regions
- IG_TR_OVERLAP: flag if hotspot overlaps immunoglobulin regions

### Warning positions
During the process of parsing of mutations some unexpected data can be observed and are reported in a separate
 TSV file, warningpositions.txt, with the following header: 
 - CHROMOSOME: chromosome 1-22, X or Y. ('chr' prefix not included)
 - POSITION: genomic coordinate
 - REF: reference nucleotide(s)
 - ALT: alternate nucleotide(s)
 - SAMPLE: sample identifier
 - WARNING: type of warning (see Mutations parsing for further information)
 - SKIP: if True, this flag indicates that the mutation has been discarded from analysis
  
# How to cite and additional information

If you use HotspotFinder, please cite our manuscript at: 


Additional information about HotspotFinder can be found at the Supplementary Note 1 of our manuscript. 
